opm/simulators/linalg/bda/WellContributionsOCLContainer.cpp

```diff
@@ -0,0 +1,113 @@
+#include <dune/common/timer.hh>
+
+#include <opm/simulators/linalg/bda/WellContributionsOCLContainer.hpp>
+#include<iostream>
+
+namespace bda
+{
+    using Opm::OpmLog;
+    using Dune::Timer;
+
+    void WellContributionsOCLContainer::init(Opm::WellContributions &wellContribs, int Nb_){
+        Nb = Nb_;
+        dim = wellContribs.dim;
+        dim_wells = wellContribs.dim_wells;
+
+        if(!wellContribs.h_val_pointers_ocl.empty()){
+            num_std_wells = wellContribs.h_val_pointers_ocl.size() - 1;
+
+            s.Cnnzs = cl::Buffer(*context, CL_MEM_READ_WRITE, sizeof(double) * wellContribs.h_Cnnzs_ocl.size());
+            s.Dnnzs = cl::Buffer(*context, CL_MEM_READ_WRITE, sizeof(double) * wellContribs.h_Dnnzs_ocl.size());
+            s.Bnnzs = cl::Buffer(*context, CL_MEM_READ_WRITE, sizeof(double) * wellContribs.h_Bnnzs_ocl.size());
+            s.Ccols = cl::Buffer(*context, CL_MEM_READ_WRITE, sizeof(int) * wellContribs.h_Ccols_ocl.size());
+            s.Bcols = cl::Buffer(*context, CL_MEM_READ_WRITE, sizeof(int) * wellContribs.h_Bcols_ocl.size());
+            s.val_pointers = cl::Buffer(*context, CL_MEM_READ_WRITE, sizeof(unsigned int) * wellContribs.h_val_pointers_ocl.size());
+            s.toOrder = cl::Buffer(*context, CL_MEM_READ_WRITE, sizeof(int) * Nb);
+        }
+        else{
+            num_std_wells = 0;
+        }
+    }
+
+    void WellContributionsOCLContainer::copy_to_gpu(Opm::WellContributions &wellContribs){
+        if(num_std_wells > 0){
+            toOrder.insert(toOrder.end(), wellContribs.toOrder, wellContribs.toOrder + Nb);
+
+            cl::Event event;
+            queue->enqueueWriteBuffer(s.Cnnzs, CL_TRUE, 0, sizeof(double) * wellContribs.h_Cnnzs_ocl.size(), wellContribs.h_Cnnzs_ocl.data());
+            queue->enqueueWriteBuffer(s.Dnnzs, CL_TRUE, 0, sizeof(double) * wellContribs.h_Dnnzs_ocl.size(), wellContribs.h_Dnnzs_ocl.data());
+            queue->enqueueWriteBuffer(s.Bnnzs, CL_TRUE, 0, sizeof(double) * wellContribs.h_Bnnzs_ocl.size(), wellContribs.h_Bnnzs_ocl.data());
+            queue->enqueueWriteBuffer(s.Ccols, CL_TRUE, 0, sizeof(int) * wellContribs.h_Ccols_ocl.size(), wellContribs.h_Ccols_ocl.data());
+            queue->enqueueWriteBuffer(s.Bcols, CL_TRUE, 0, sizeof(int) * wellContribs.h_Bcols_ocl.size(), wellContribs.h_Bcols_ocl.data());
+            queue->enqueueWriteBuffer(s.val_pointers, CL_TRUE, 0, sizeof(unsigned int) * wellContribs.h_val_pointers_ocl.size(), wellContribs.h_val_pointers_ocl.data());
+            queue->enqueueWriteBuffer(s.toOrder, CL_TRUE, 0, sizeof(int) * toOrder.size(), toOrder.data(), nullptr, &event);
+            event.wait();
+        }
+    }
+
+    void WellContributionsOCLContainer::update_on_gpu(Opm::WellContributions &wellContribs){
+        if(num_std_wells > 0){
+            cl::Event event;
+            queue->enqueueWriteBuffer(s.Cnnzs, CL_TRUE, 0, sizeof(double) * wellContribs.h_Cnnzs_ocl.size(), wellContribs.h_Cnnzs_ocl.data());
```

Comment:
Won't work in general. We only treat open wells and the number of these might change. Even for each wellsthe number of cells perforated might change (different sparsity pattern). Hence we cannot reuse these buffers as their size might be inadequate and the column information wrong.

Reply:
Seems to be still an issue. If you disagree then please point to where the size if the device buffers is changed or explain why this is not needed (I might be missing something).

Reply:
This is strange. I've implemented a function called `reinit` that recreates the buffers if the vectors change size. I don't know why it is not showing here. 

There is an "outdated" flag appearing in this conversation, maybe this is why.

But the change is commited, I've checked in my repository.

Reply:
Must have missed that. It is here, too.